### DISCUSSION TITLE
Poor performance in "No UDP" mode

### DISCUSSION THREAD

**Discussion by bizzbyster** [vpnhood/VpnHood#98] [Q&A]
I ran a test comparing VpnHood to Squid installed on the same Linux server, and _consistently_ see much better speedtest results with Squid against a Speedtest server in the same region as the Linux server.

![image](https://user-images.githubusercontent.com/4001792/153432059-8523e8c8-a7f9-45b5-8121-9742971811ef.png)

when my browser was connected to squid (see results [here](https://www.speedtest.net/result/12742631819 )) and...

![image](https://user-images.githubusercontent.com/4001792/153432254-5268ea5c-31ae-4bd7-85c0-5b24a1eb29f8.png)
when VpnHood was connected using the same laptop and web browser (see results [here](https://www.speedtest.net/result/12742612956)).

Note that I am on the East Coast and my Linux server (running VpnHood server and squid) is on the West Coast b/c I wanted to add some latency to my connection.

Any ideas why this might be happening?

**Comment by trudyhood** [comment]
Dear @bizzbyster Thank you for your contribution and interest in VpnHood
It shows that there is much more room for optimization, but I think we will never reach the speed of Squid because:

- **Squid**: **App -> SquidServer (SOCKS proxy) -> Target Site**
- **VpnHood**: **App -> PacketFilter -> Local VpnHood Proxy -> VpnHood Server -> Target Site**

PacketFilter usually impacts much on performance and battery usage. Of course, if we add SOCKS proxy interface to VpnHood, then users can skip PacketFilter, but they need to set and use a proxy for their App or device. In this case, advanced users can split tunneling by App too and huge battery saving on mobile devices.

- **VpnHood**: **App -> Local VpnHood Proxy (SOCKS proxy) -> VpnHood Server -> Target Site**

We can also add (SOCKS proxy) support to the VpnHood server, but in this case, VpnHood will be the same as Squid but in C# (looks reinventing the wheel). Also, it may have a SOCKS proxy fingerprint and can be detected in this case which I don't know more about it.

- **VpnHood**: **App -> VpnHood Server (SOCKS proxy) -> Target Site**

**Comment by bizzbyster** [comment]
I believe you disable Nagle in the TCP session between client and server. Is there some way I can re-enable that to see if it helps? I have trouble believing that the problem is caused by PacketFilter.

**Reply by trudyhood** [reply]
You need to recompile the code.

We put many tests over it and found out the performance did not change. We ended up disabling it for two reasons:
* Of course, it should be disabled for TcpDatagram as single packets must go out as soon as possible
* For TcpProxy sessions: If you enable it, you will suppress any app configuration on TCP Nagle, also the app TCP stack has already been delayed and optimized the incoming data, so as a proxy, enabling the Nagle algorithm does not have benefits.

**Reply by trudyhood** [reply]
WinDivert PacketFilter hugely decreases the performance from 500 Mbit/s to 180 MBit/s unless we use batch mode. you can test it purely netfilter on WinDivert. VpnHood uses batch mode, but we use sharppcap as the .net wrapper and we are waiting for them. However, it shouldn't impact the performance unless you have a slow machine. see the following issue
https://github.com/dotpcap/sharppcap/issues/273

Android VpnService, generally causes a huge impact on performance and battery usage. What could we do is do our best at the application level. The only remedy should be SOCKS proxy.

**Reply by bizzbyster** [reply]
Thanks for all this detailed info!

**Reply by bizzbyster** [reply]
Can I test increasing the number of TcpDatagramChannels in the pool? That might contribute to the issue.

**Reply by trudyhood** [reply]
In client, you need to change it in code, but I can add it to its settings.json if you think it would be useful. We believed that is for diagnostic and haven't added it to settings. You can find it here: https://github.com/vpnhood/VpnHood/blob/26fb52c27e407cd845e3017570a845198960ec80/VpnHood.Client/ClientOptions.cs
**Client MaxDatagramChannelCount default value is 4**

You need to increase MaxDatagramChannelCount in the server too, otherwise, the server will close the old connection.
**Server MaxDatagramChannelCount default value is 8**
```
{
  "RestAccessServer": null,
  "FileAccessServer": {
    "TcpEndPoints": [ "0.0.0.0:443", "[::]:443" ],
    "SslCertificatesPassword": null,
    "TrackingOptions": {
      "LocalPort": true,
      "ClientIp": true
    },
    "SessionOptions": {
      "Timeout": 600,
      "UdpTimeout": 60,
      "TcpTimeout": 900,
      "IcmpTimeout": 30,
      "MaxDatagramChannelCount": 8,
      "MaxUdpPortCount": 0,
      "TcpBufferSize": 8192
    }
  },
  "IsDiagnoseMode": false
}
```

**Reply by bizzbyster** [reply]
I should learn to compile the client and server in any case. 

Follow up question: does a TCP connection from the application have a stickiness to a single TcpDatagramChannel? So, for instance, if I run a speedtest.net, which uses 4 connections by default as I understand it, then each of those TCP connections should be handled by a single TcpDatagramChannel. Is that right? Thanks again for all this information!

**Reply by trudyhood** [reply]
No, TcpDatagram channels are useless for most apps.

In your scenario
1. You type speedtest.net so your browser request the IP address of speedtest.net by DNS server, sending a UDP packet that is handled via TcpDatagramChannel (it is just a single UDP packet, so only once of TcpDatagramChannel is adequate)
2. You press start test then speedtest.net try to establish one or a few TCP connections (Depends on your settings)
![image](https://user-images.githubusercontent.com/72891124/153954888-57aa839c-8d82-4326-b5bb-8089cb3da5db.png)

In either case, speedtest.net tries to download some data from servers. All of these are TCP, so VpnHood uses TcpProxyChannel, not TcpDatagramChannel. It means the TcpDatagramChannel is not used in most cases unless an app uses UDP. VpnHood establishes as many as TcpProxyChannels that are requested.

**Reply by bizzbyster** [reply]
Ok now I get it thank you!
